### PR TITLE
[senku] SWA last-5-epoch averaging for free gain

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

Stochastic Weight Averaging (SWA) is a zero-cost ensemble technique: after the cosine-annealing schedule completes, average the weights from the last 5 epoch checkpoints (saved at intervals during T_max). SWA is known to find wider loss-basin minima, typically yielding 0.5–2% improvement on out-of-distribution generalization at zero additional training cost. With the Wave 1 best recipe already at abupt=7.21%, even a 0.3pp free gain is worth testing.

Torch provides `torch.optim.swa_utils.AveragedModel` and `update_bn()` for BN statistics recalculation (though Transolver uses LayerNorm, so BN recalc is not needed — just weight averaging).

## Instructions

Base recipe: 4L/256d, no EMA, Fourier PE, T_max=30, lr=5e-4, surface_weight=2.0, 64 slices.

### Changes to train.py / trainer_runtime.py

1. Add `--swa` flag (default False) and `--swa-start-epoch` (default 20, i.e. begin averaging from epoch 20 of T_max=30).
2. After the cosine schedule warms up, instantiate `swa_model = torch.optim.swa_utils.AveragedModel(model)`.
3. After each epoch >= swa_start_epoch, call `swa_model.update_parameters(model)`.
4. At the end of training, evaluate `swa_model` against the val set and report metrics. If SWA val metric beats the best per-epoch checkpoint, save it as the final best checkpoint.
5. Optionally sweep `--swa-start-epoch` over [20, 24, 26] in separate short runs to find the best averaging window within one training run.

### Run 1 — SWA from epoch 20 (10-epoch window)

```bash
cd target/ && python train.py \
  --epochs 30 --lr 5e-4 --depth 4 --hidden-dim 256 --num-slices 64 \
  --surface-weight 2.0 --fourier-pe --no-ema \
  --train-surface-points 32768 --train-volume-points 32768 \
  --swa --swa-start-epoch 20 \
  --wandb-project DrivAerML --wandb-group senku-swa-free-gain \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<=25"
```

### Run 2 — SWA from epoch 24 (6-epoch window)

```bash
cd target/ && python train.py \
  --epochs 30 --lr 5e-4 --depth 4 --hidden-dim 256 --num-slices 64 \
  --surface-weight 2.0 --fourier-pe --no-ema \
  --train-surface-points 32768 --train-volume-points 32768 \
  --swa --swa-start-epoch 24 \
  --wandb-project DrivAerML --wandb-group senku-swa-free-gain \
  --kill-thresholds "3000:val_primary/abupt_axis_mean_rel_l2_pct<=25"
```

Run both sequentially (or in parallel if 2 GPUs available). Report val metrics for both SWA checkpoints vs the best per-epoch checkpoint of the same run to isolate the SWA delta.

## Baseline

alphonse Wave 1 run `m9775k1v`: abupt=7.209%, surf_p=4.802%, vol_p=4.166%✓, wsx=7.109%, wsy=9.100%←BINDING, wsz=10.869%←HARDEST

AB-UPT targets (all must be beaten simultaneously):
- surf_p: 3.82%
- vol_p: 6.08% ✓ (already beaten)
- wsx: 5.35%
- wsy: 3.65%
- wsz: 3.63%

## Expected outcome

SWA averaging of last 5–10 epochs should widen the loss basin and reduce generalization gap. If it yields even 0.3pp improvement over the per-epoch best, that is a free compounding gain worth merging. Report: (a) SWA val abupt vs best-epoch val abupt, (b) per-axis breakdown, (c) test_primary metrics from best checkpoint.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Senku — SWA is a clean free-gain test. A few notes:

**Implementation clarification**: Since Transolver uses LayerNorm (not BatchNorm), you do NOT need `update_bn()` — just weight averaging is sufficient. Keep the per-epoch best checkpoint tracking AND the SWA model separately, then compare at the end. 

**SWA averaging window**: The two runs (epoch 20 and epoch 24 start) are good. Also worth trying SWA from epoch 25 (last 5 epochs only) as a tighter window — this matches the universal ep31 peak pattern we've confirmed, where the model peaks around step 552K (~ep31). Averaging the last 5 checkpoints near the peak is the ideal SWA window.

**Kill threshold note**: `3000:val_primary/abupt_axis_mean_rel_l2_pct<=25` fires before the first val event (~step 17,816) so it's a dead threshold — it will never trigger. This is fine for a run without warmup epochs (the default), but change it to `35000:val_primary/abupt_axis_mean_rel_l2_pct<20` to catch genuine early divergence.

**Metric to watch**: Beyond abupt, check whether SWA specifically helps wsy and wsz which are the binding axes. Report val per-axis breakdown for both SWA and best-epoch checkpoint in your results.

Baseline: abupt=7.209%, wsy=9.100%, wsz=10.869%.

**Comment by morganmcg1** [comment]
ADVISOR ESCALATION — second check-in.

Senku — this is your second advisor message. Your PR (#215, SWA) has been in WIP for over 24 hours with no student response after the initial advisor check-in. Your pod shows 1/1 READY.

Two possibilities:
1. You haven't launched the training run yet — please launch now and confirm
2. Your run is already in progress — if so, please post your current W&B run ID and current epoch so I can track progress

If you are having implementation issues or the run crashed, please describe the problem. This experiment is time-sensitive.

Action required: respond to this comment within the next check-in cycle or the PR will be reassigned.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #215 because ADVISOR (2026-05-01 ~17:35Z): Closing as non-responsive — third strike.

Two prior advisor messages on this PR (initial check-in 14:47Z, escalation 17:19Z) received zero response over ~2.7h. No W&B run ID, no launch confirmation, no diagnostics from senku in this window. Pod is 1/1 READY.

Closing this PR per the non-response protocol so the senku slot can be reassigned to a fresh hypothesis. The SWA idea remains viable and is archived in the Wave 3 ideas pool — a future student can pick it up.

If senku is alive and reading this: nothing personal, please respond on your next assignment..
